servo/pub_sub.py

```diff
@@ -0,0 +1,430 @@
+
+    Channel names conform to [RFC 1123](https://tools.ietf.org/html/rfc1123) and must:
+        * contain no more than 253 characters
+        * contain only lowercase alphanumeric characters, '-' or '.'
+        * start with an alphanumeric character
+        * end with an alphanumeric character
+
+    Attributes:
+        name: The unique name of the Channel within the Exchange.
+        description: An optional supplemental description of the Channel.
+        created_at: The date and time that the Channel was created.
+        exchange: The pub/sub Exchange that the Channel belongs
+    """
+    name: ChannelName
+    description: Optional[str] = None
+    created_at: datetime.datetime = pydantic.Field(default_factory=datetime.datetime.now)
+    exchange: Exchange
+
+    async def publish(self, message: Message) -> None:
+        """Publish a Message into the Channel."""
+        await self.exchange.publish(message, self)
+
+
+class Publisher(pydantic.BaseModel):
+    """A Publisher broadcasts Messages to a Channel in an Exchange.
+
+    Publishers are asynchronously callable to publish a Message.
+
+    Attributes:
+        exchange: The pub/sub Exchange that the Publisher belongs to.
+        channel: The Channel that the Publisher publishes Messages to.
+    """
+    exchange: Exchange
+    channel: Channel
+
+    async def __call__(self, message: Message) -> None:
+        await self.exchange.publish(message, self.channel)
+
+
+class Exchange(pydantic.BaseModel):
+    """An Exchange facilitates the publication and subscription of Messages in Channels."""
+    channels: Set[Channel] = {}
+    _publishers: Set[Subscriber] = pydantic.PrivateAttr({})
+    _subscribers: Set[Subscriber] = pydantic.PrivateAttr({})
+    _queue: asyncio.Queue = pydantic.PrivateAttr(default_factory=asyncio.Queue)
+    _queue_processor: asyncio.Task = pydantic.PrivateAttr()
+
+    def start(self) -> None:
+        """Start exchanging Messages between Publishers and Subscribers."""
+        if self.is_running:
+            raise RuntimeError("the Exchange is already running")
+        self._queue_processor = asyncio.create_task(self._process_queue())
+
+    async def shutdown(self) -> None:
+        """Shutdown the Exchange by processing all Messages and clearing all child objects."""
+        if not self.is_running:
+            raise RuntimeError("the Exchange is not running")
+        await self._queue.join()
+        self._queue_processor.cancel()
+        await asyncio.gather(self._queue_processor, return_exceptions=True)
+
+        # Purge all our children to break object cycles
+        self.channels.clear()
+        self._publishers.clear()
+        self._subscribers.clear()
+
+    async def _process_queue(self) -> None:
+        while True:
+            message, channel = await self._queue.get()
+            if message is None:
+                # Exit condition
+                break
+
+            results = asyncio.gather(
```

Comment:
I assume you run one exchange for all the servox message queue. Otherwise if one exchange per channel, there is no need to use the channel any more.

Here it has the issue of serializing all the message and caller. For example, let's say there are two channel of message is totally unrelated. Let's suppose channel "a" has message "A" then channel "b" has message "B". 

The processing will force message "B" does not start to get process before *all* of the call back of the message "A" completes. The is not optimal because "b" channel might have nothing to do with "a". That means any process handler can stall the processing pipe which can be a design draw back.

Reply:
There will be one `Exchange` shared across all Connectors via the mixin -- just like the existing event bus.

You are correct that there is a serialization of message processing here which is really a convenience for handling the exceptions. There are other approaches that can yield more concurrency with a bit more leg-work. Noted for review